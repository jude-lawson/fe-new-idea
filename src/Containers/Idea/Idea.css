.idea-container {
  align-items: center;
  display: flex;
  flex-direction: column;
  justify-content: center;
  padding: 2rem;
  color: var(--primary-dark-grey);
  background: #ffba49;
  margin: 8rem auto 2rem;
  box-shadow: 8px 10px 30px 0 rgba(46, 61, 73, 0.15);
  max-width: 116rem;
}

.idea {
  padding: 0 4rem;
}

.idea-button--container {
  display: flex;
  justify-content: space-around;
  align-items: center;
  width: 100%;
}

.idea-button--container .btn-grad {
  padding: 1rem 1.5rem;
}

.idea-button--container .btn-grad a {
  text-decoration: none;
  color: var(--primary-white);
}

.idea-button--container .blue-btn {
  background-image: linear-gradient(to right, #289dc5, #50b4be);
  color: var(--primary-white);
}

.idea-author {
  display: flex;
  align-items: center;
  font-size: 2.6rem;
  margin-bottom: 2rem;
}

.idea-author span {
  margin-right: 2rem;
}

.idea-image img {
  max-height: 5rem;
  border-radius: 50%;
  background: #fff;
}

.idea-title {
  font-size: 3rem;
  margin-bottom: 2rem;
}

.idea-content {
  font-size: 2rem;
  margin-bottom: 1rem;
}

.loader-wrapper {
  align-items: center;
  display: flex;
  height: 100vh;
  justify-content: center;
}

.scale-up-hor-left {
  animation: scale-up-hor-left 0.4s cubic-bezier(0.39, 0.575, 0.565, 1) both;
}

/* ----------------------------------------------
 * Generated by Animista on 2018-9-18 12:13:29
 * w: http://animista.net, t: @cssanimista
 * ---------------------------------------------- */

/**
 * ----------------------------------------
 * animation scale-up-hor-left
 * ----------------------------------------
 */
@keyframes scale-up-hor-left {
  0% {
    transform: scaleX(0.4);
    transform-origin: 0% 0%;
  }

  100% {
    transform: scaleX(1);
    transform-origin: 0% 0%;
  }
}
