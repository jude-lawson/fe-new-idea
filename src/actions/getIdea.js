export const getIdeaAction = idea => ({
  type: 'GET_IDEA',
  idea
});