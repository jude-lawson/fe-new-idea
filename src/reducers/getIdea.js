export const getIdeaReducer = (state = {}, action) =>  {
  switch (action.type) {
    case 'GET_IDEA':
      return {
        ...action.idea
      };
    default:
      return state;
  }
};