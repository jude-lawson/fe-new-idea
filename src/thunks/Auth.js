import { userLoginAction, userLogOutAction } from '../actions/auth';
import { githubOAuthLogin, logout } from '../firebase/firebase';

export const firebaseLogin = () => {
  return async dispatch => {
    const newUser = await githubOAuthLogin();
    dispatch(userLoginAction(newUser));
  };
};

export const firebaseLogout = () => {
  return async dispatch => {
    await logout();
    dispatch(userLogOutAction());
  };
};