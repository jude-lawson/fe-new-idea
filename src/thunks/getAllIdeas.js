import { getAllIdeasAction } from '../actions/ideasAll';
import { getIdeas } from '../api-calls/api-calls';

export const getAllIdeas = () => {
  return async dispatch => {
    const ideas = await getIdeas();
    dispatch(getAllIdeasAction(ideas));
  };
};