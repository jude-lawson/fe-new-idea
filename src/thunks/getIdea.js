import { getIdeaAction } from '../actions/getIdea';
import { getIdea } from '../api-calls/api-calls';

export const getIdeaById = id => {
  return async dispatch => {
    const idea = await getIdea(id);
    dispatch(getIdeaAction(idea));
  };
};