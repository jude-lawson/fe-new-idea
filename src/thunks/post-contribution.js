import { addComment } from '../actions/comment';

export const postContribution = comment => {
  return async dispatch => {
    const contribution = await jest.fn();
    console.log(contribution);
    // const comment = {...comment, ...contribution }
    dispatch(addComment(comment));
  };
};