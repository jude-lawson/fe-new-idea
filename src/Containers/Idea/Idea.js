import React, { Component } from 'react';
import { PropTypes } from 'prop-types';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import Contributions from '../../Components/Contributions/Contributions';
import { getIdeaById } from '../../thunks/getIdea';
import { GridLoader } from 'halogenium';
import placeholder from '../../images/place-holder.png';
import { history } from '../../router/Router';

import './Idea.css';

export class Idea extends Component {
  constructor() {
    super();
    this.state = {
      loading: true,
      id: window.location.pathname.replace(/\D+/g, ''),
      contributions: []
    };
  }

  componentDidMount = async () => {
    this.setState({ loading: true });

    await this.getIdeaByIdFromDb();
    this.getContributionsById();
    this.setState({ loading: false });
  }

  getIdeaByIdFromDb = async () => {
    const id = this.state.id;
    await this.props.getIdeaById(id);
  }

  getContributionsById = () => {
    const contributions = this.props.idea.contributions;
    this.setState({ contributions });
  }

  render() {
    const { loading, contributions } = this.state;
    const { idea } = this.props;

    if (loading) {
      return (
        <div className="app-container">
          <div className="loader-wrapper">
            <GridLoader />
          </div>
        </div>
      );
    }

    return (
      <section className="app-container">
        {idea &&
        <React.Fragment>
          <article className="idea-container">
            <div className="idea-button--container">
              <button className="btn-grad blue-btn" onClick={() => history.goBack()} >&larr; Back</button>
              <button className="btn-grad btn-small blue-btn"><Link to="/ideaform">New Idea &rarr;</Link></button>
            </div>
            <div className="idea">
              <h2 className="idea-title">{idea.title}</h2>
              <p className="idea-author">
                <span className="idea-image">
                  <img
                    src={idea.author.profile_pic_url ? idea.author.profile_pic_url : placeholder}
                    alt={idea.author.username}/>
                </span>
                {idea.author ? idea.author.username : null}
              </p>
              <p className="idea-content">{idea.body}</p>
            </div>
          </article>
          <Contributions contributions={contributions} />
        </React.Fragment>
        }
      </section>
    );
  }
}

Idea.propTypes = {
  idea: PropTypes.object,
  getIdeaById: PropTypes.func
};

export const mapStateToProps = state => ({
  idea: state.idea
});

export const mapDispatchToProps = dispatch => ({
  getIdeaById: id => dispatch(getIdeaById(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(Idea);
