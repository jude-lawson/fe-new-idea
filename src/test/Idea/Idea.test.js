import React from 'react';
import { shallow } from 'enzyme';
import { Idea, mapStateToProps, mapDispatchToProps } from '../../Containers/Idea/Idea';
import { getIdeaById } from '../../thunks/getIdea';

describe('<Idea />', () => {
  let wrapper;
  const mockIdea ={
    title: 'Hello',
    body: 'World',
    author: {
      username: 'defaultuser'
    },
    contributions: [
      {
        id: 2,
        title: 'I love hello worlds',
        body: 'so much love'
      }
    ]
  };
  const mockGetIdeaById = jest.fn();

  beforeEach(() => wrapper = shallow(<Idea getIdeaById={mockGetIdeaById} />));

  test('should call getIdeaByIdFromDb on componentDidMount', () => {
    const spy = jest.spyOn(wrapper.instance(), 'getIdeaByIdFromDb');
    wrapper.instance().componentDidMount();
    expect(spy).toHaveBeenCalled();
  });

  test('should call getIdeaById', () => {
    const spy = wrapper.instance().props.getIdeaById;
    wrapper.instance().componentDidMount();
    expect(spy).toHaveBeenCalled();
  });

  describe('mapStateToProps', () => {
    test('should map idea to to state', () => {
      const mockState = {
        idea: mockIdea
      };

      const expectedState = {
        idea: mockIdea
      };

      const mappedProps = mapStateToProps(mockState);

      expect(mappedProps).toEqual(expectedState);
    });
  });

  describe('mapDispatchToProps', () => {
    test('should getIdeaById to props', () => {
      const mockId = 3;
      const mockDispatch = jest.fn();
      const actionToDispatch = getIdeaById(mockId);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.getIdeaById();

      expect(mockDispatch).toHaveBeenCalled();
    })
  })


  test('should render when idea present', () => {
    wrapper = shallow(<Idea idea={mockIdea} />)
  });

  test('should render loading when loading is true', () => {
    wrapper.setState({ loading: true });
    expect(wrapper).toMatchSnapshot();
  });

  test('should render without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
