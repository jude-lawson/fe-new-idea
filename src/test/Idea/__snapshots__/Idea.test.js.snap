// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Idea /> should render loading when loading is true 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Idea
    getIdeaById={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Connect(Header) />,
        <div
          className="idea-container"
>
          <h1>
                    loading
          </h1>
</div>,
      ],
      "className": "app-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <h1>
            loading
</h1>,
          "className": "idea-container",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "loading",
          },
          "ref": null,
          "rendered": "loading",
          "type": "h1",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Connect(Header) />,
          <div
            className="idea-container"
>
            <h1>
                        loading
            </h1>
</div>,
        ],
        "className": "app-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <h1>
              loading
</h1>,
            "className": "idea-container",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "loading",
            },
            "ref": null,
            "rendered": "loading",
            "type": "h1",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "disableLifecycleMethods": true,
    "hydrateIn": undefined,
  },
}
`;

exports[`<Idea /> should render without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Idea
    getIdeaById={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Connect(Header) />,
        undefined,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      undefined,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Connect(Header) />,
          undefined,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        undefined,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "disableLifecycleMethods": true,
    "hydrateIn": undefined,
  },
}
`;
